clients/src/main/java/org/apache/kafka/common/errors/FatalExitError.java

```diff
@@ -0,0 +1,77 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.common.errors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The exception that indicates the need to exit the process. This exception is expected to be
+ * caught at the highest level of the thread so that no shared lock would be held by the thread
+ * when it calls {@link System#exit(int)}
+ */
+public class FatalExitError extends Error {
+
+    private static final Logger log = LoggerFactory.getLogger(FatalExitError.class);
+
+    private final static long serialVersionUID = 1L;
+
+    protected int exitStatus;
+
+    public FatalExitError(int exitStatus, String message, Throwable cause) {
+        super(message, cause);
+        this.exitStatus = exitStatus;
+    }
+
+    public FatalExitError(int exitStatus, String message) {
+        super(message);
+        this.exitStatus = exitStatus;
+    }
+
+    public FatalExitError(int exitStatus, Throwable cause) {
+        super(cause);
+        this.exitStatus = exitStatus;
+    }
+
+    public FatalExitError(int exitStatus) {
+        super();
+        this.exitStatus = exitStatus;
+    }
+
+    public void shutdownSystem() {
```

Comment:
Thanks @ijuma I am still not clear on the suggested approach. Let me try to clarify my understanding:

1. A Kafka thread, `ft`, faces a fatal error
2. It starts an anonymous thread, `at`, which calls `System.exit`, which starts the shutdown thread, `st`.
3. `ft` itself has to exit otherwise we would have a deadlock
4. `st` will invoke `KafkaServer.shutdown`
4. After `KafkaServer.shutdown` is finished any main thread waiting on `awaitShutdown` will also resume and finish. At this point the main thread would not call `System.exit` again.

Step 3 is achieved by using FatalExitError and making sure that no place in the code it is caught and swallowed.
Step 2 can occur at any point in time, i) right after the fatal error, ii) some place in the thread after catching the FatalExitError, etc. My suggestion is to execute that right away.

So can you explain that which of the steps above are targeted by the suggested approach and how it would be changed.